test/sequential/test-net-server-bind.js

```diff
@@ -62,3 +62,185 @@ const net = require('net');
+      assert.strictEqual(address, '::1');
+      assert.strictEqual(listenPort, testPort);
+
+      server.close();
+    }));
+  }
+
+  // TODO host with invalid port throws
+  // currently the URL constructor throws this error
+  // common.expectsError(() => {
+  //   net.createServer.listen({
+  //     host: '127.0.0.1:99999999999999',
+  //   }, common.mustNotCall());
+  // }, {
+  //   code: 'ERR_SOCKET_BAD_POT',
+  //   type: RangeError,
+  // });
+
+  // Host without port AND without port option throws
+  common.expectsError(() => {
+    net.createServer().listen({
+      host: '127.0.0.1',
+    }, common.mustNotCall());
+  }, {
+    code: 'ERR_INVALID_ARG_VALUE',
+    type: TypeError,
+    message: /^The argument 'options' must have the property "port" or "path"\. Received .+$/,
+  });
+
+  // Host with port and port option
+  {
+    const testPort = ++port;
+    const host = `127.0.0.1:${port}`;
+
+    const server = net.createServer();
+    server.listen({ host, port: testPort }, common.mustCall(() => {
+      const { address, port: listenPort } = server.address();
+      assert.strictEqual(address, '127.0.0.1');
+      assert.strictEqual(listenPort, testPort);
+
+      server.close();
+    }));
+  }
+
+  // TODO host with port AND port option *as a string*
+  // This errors out as converting the port to a number happens really
+  // late, only after we parse the host option
+  // {
+  //   const testPort = ++port;
+  //   const host = `127.0.0.1:${port}`;
+
+  //   const server = net.createServer();
+  //   server.listen({ host, port: String(testPort) }, common.mustCall(() => {
+  //     const { address, port: listenPort } = server.address();
+  //     assert.strictEqual(address, '127.0.0.1');
+  //     assert.strictEqual(listenPort, testPort);
+
+  //     server.close();
+  //   }));
+  // }
+
+  // Host with port and a conflicting port option throws
+  {
+    common.expectsError(() => {
+      net.createServer().listen({
+        host: '127.0.0.1:1234',
+        port: 4321,
+      }, common.mustNotCall());
+    }, {
+      type: TypeError,
+      message: /^Mismatched port in "host" and "port" options$/,
+    });
+  }
+
+  // Host with with hostname option
+  {
+    const server = net.createServer();
+    server.listen({
+      host: '127.0.0.1',
+      hostname: '127.0.0.1',
+      port: 0,
+    }, common.mustCall(() => {
+      const { address } = server.address();
+      assert.strictEqual(address, '127.0.0.1');
+
+      server.close();
+    }));
+  }
+
+  // Host with hostname and a conflicting hostname option throws
+  {
+    common.expectsError(() => {
+      net.createServer().listen({
+        host: '127.0.0.1',
+        hostname: '127.1.1.0',
+        port: 0,
+      }, common.mustNotCall());
+    }, {
+      type: TypeError,
+      message: /^Mismatched hostname in "host" and "hostname" options$/,
+    });
+  }
+
+  // A parsed URL
+  // TODO this currently errors out, see the port-as-string mentioned above
```

Comment:
This is still a WIP right..?

Reply:
Yes :) I've mentioned this line in the PR comment as well.

This commit is a draft to show the overall intention. If deemed interesting enough these kinds of things will be ironed out - exactly what and how is outlined in the PR comment.